### PR TITLE
chore: update GH Actions workflows

### PR DESCRIPTION
**Pull request by felix-seifert**
This PR is meant to upgrade the different GH Actions workflows to use new versions of existing actions and use new functionalities of these actions.

Changes other than formatting:
* Update `actions/checkout` to version `4`
* Update `actions/setup-java` to version `3`
* Use cache for Maven
* User Amazon Corretto JDK distribution
* Ignore being triggered by changes in `.gitignore` and `*.md` files
* Remove usage of already archived `samuelmeuli/action-maven-publish@v1`
* Perform same action as `samuelmeuli/action-maven-publish@v1` by adding necessary release credentials to `actions/setup-java@v3` and calling `mvn deploy` manually

I tested the changed workflow triggered by `pull_request` locally. Due to the needed secrets for the `push` workflow, I could not test the releasing properly.

### CODE REVIEW

File: .github/workflows/pullrequest.yml
Comment (dennisgranath) [inline .github/workflows/pullrequest.yml:8]: Would the necessary checks be green if there is only a change in the readme file?

**Review by dennisgranath — APPROVED** [review]
LGTM, I guess we notices if release doesn't work after merge...

File: .github/workflows/pullrequest.yml
Comment (felix-seifert) [reply .github/workflows/pullrequest.yml:8]: Sounded like a basic question but is indeed an interesting issue. According to [the GitHub docs](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/collaborating-on-repositories-with-code-quality-features/troubleshooting-required-status-checks#handling-skipped-but-required-checks), if a certain check is required and will be skipped because of some path rules, the check remains in a pending state. The branch `master` is protected and requires the check `build` to succeed. This means that a PR which changes only `.md` files cannot become green.

We should therefore ignore this optimisation of skipping certain paths for the `pullrequest` workflow. The `release` workflow can still ignore certain paths because it does not contain any required checks. I will commit the necessary changes.

**Review by dennisgranath — APPROVED** [review]

**Pull request merged**
